### PR TITLE
feat(native_filter_migration): add transition mode

### PR DESCRIPTION
**Pull request by graceguo-supercat**
### SUMMARY
As part of Superset roadmap and [SIP-64](https://github.com/apache/superset/issues/14383), by the end of this year 2021, Superset will not support filter_box anymore. This PR is to create a transition mode for dashboards with **only** filter_box (and without filter component):  when feature flag `ENABLE_FILTER_BOX_MIGRATION` is enabled,  dashboard will automatically convert filter_box to dashboard native filter component in front-end only, so that user can review the converted dashboard, then save or abandon change in case any issues found. 

This PR is mostly designed for airbnb users who used dashboard for many years, and had large dashboards with many charts, tabs, and scoped filters. Manually convert filter_box to filter component for us will be a nightmare. This PR is to make sure our heavy dashboards can go through a smooth transition to the next state. For dashboards that have mixed filter_box and filter components, we assume the size of dashboard is pretty small and the number of filter_box must be manageable, the dashboard owner should convert their filter_box manually.

At the end of transition time, we will run db migration which hard converts all filter_box into filter component in dashboard. 

**Note**:  The major goal for this PR is to define **_different dashboard and filter_box behavior_** during transition time. I may need to add a few small UI change PR to improve, but don't want to make this PR too big to review.

During transition time, this is dashboard and filter_box status under different state:

**Existing dashboard:**

v3

**Creating new dashboard:**

Screen Shot 2021-10-06 at 8 39 25 AM

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
**Before**
Bu9nvdXeAn

**After**

- For users having filter_box and enabled `ENABLE_FILTER_BOX_MIGRATION` feature flag:

ozricAq2gC

- For users having filter_box but without `ENABLE_FILTER_BOX_MIGRATION` feature flag:

Screen Shot 2021-10-08 at 11 43 38 AM

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [x] Required feature flags:
- [x] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/src/common/hooks/apiResources/dashboards.ts [hunk 1]
```diff
@@ -17,19 +17,33 @@
...
+      try {
+        // need handle data parsing error
+        if (dashboard.json_metadata) {
+          metadata = JSON.parse(dashboard.json_metadata);
+        }
+        if (dashboard.position_json) {
+          position_data = JSON.parse(dashboard.position_json);
+        }
+      } catch (e) {
+        console.error('can not parse dashboard metadata.');
```
[the remark this answers is not in the document]
Comment (graceguo-supercat) [reply superset-frontend/src/common/hooks/apiResources/dashboards.ts:39]: yes! catch exception in userTransformedResource is more cleaner and covers more cases.

File: superset-frontend/src/components/Modal/Modal.tsx [hunk 2]
```diff
@@ -51,6 +51,7 @@ export interface ModalProps {
   centered?: boolean;
   footer?: React.ReactNode;
   wrapProps?: object;
+  maskStyle?: object;
```
[the remark this answers is not in the document]
Comment (graceguo-supercat) [reply superset-frontend/src/components/Modal/Modal.tsx:54]: ok. i will remove this 50px change to be consistent with all other Modals components.

**Comment by graceguo-supercat** [comment]
@kgabryje @michael-s-molina Just rebased with master and added fixes for 2nd review. Thanks!

File: superset-frontend/src/dashboard/components/DashboardBuilder/state.ts [hunk 3]
```diff
@@ -75,11 +77,12 @@ export const useNativeFilters = () => {
     if (
       filterValues.length === 0 &&
       dashboardFiltersOpen &&
-      nativeFiltersEnabled
+      nativeFiltersEnabled &&
+      ['CONVERTED', 'REVIEWING', 'NOOP'].includes(filterboxMigrationState)
```
Comment (graceguo-supercat) [inline superset-frontend/src/dashboard/components/DashboardBuilder/state.ts:81]: this line fixed the dashboard left margin issue.
this is very tricky bug. I can reproduce this issue only when:
0. enable feature flag,
1. open dashboard list page,
2. click a dashboard.

if i open dashboard from its own dashboard link, can't see this left-margin problem :joy:

File: superset-frontend/src/dashboard/components/FilterBoxMigrationModal.tsx [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    flex: 1 1 auto;
+    overflow: auto;
+  }
+
+  .modal-footer {
+    flex: 0 1 auto;
+  }
+
+  .ant-modal-body {
+    overflow: visible;
```
Comment (michael-s-molina) [inline superset-frontend/src/dashboard/components/FilterBoxMigrationModal.tsx:47]:
```suggestion
    overflow: auto;
```

**Review by michael-s-molina — COMMENTED** [review]
@graceguo-supercat Thanks for the latest updates. @jinghua-qa We are ready to test the module with the feature flag disabled and check that there are no regressions. The tests with the feature flag enabled will be made by Airbnb.

**Comment by graceguo-supercat** [comment]
Thanks!  @jinghua-qa  i will focus my tests with enabled feature flag `ENABLE_FILTER_BOX_MIGRATION`.

**Comment by jinghua-qa** [comment]
Sure, i will post the test result once i finished.

**Comment by michael-s-molina** [comment]
/testenv up FEATURE_DASHBOARD_NATIVE_FILTERS=true FEATURE_DASHBOARD_CROSS_FILTERS=true FEATURE_DASHBOARD_NATIVE_FILTERS_SET=true FEATURE_DASHBOARD_FILTERS_EXPERIMENTAL=true

**Comment by graceguo-supercat** [comment]
@michael-s-molina @jinghua-qa i found a small issue during my test (still about dashboard left margin) and pushed extra fix. You probably have to test with latest build. thanks!

**Comment by michael-s-molina** [comment]
/testenv up FEATURE_DASHBOARD_NATIVE_FILTERS=true FEATURE_DASHBOARD_CROSS_FILTERS=true FEATURE_DASHBOARD_NATIVE_FILTERS_SET=true FEATURE_DASHBOARD_FILTERS_EXPERIMENTAL=true

**Comment by jinghua-qa** [comment]
I have completed the Native Filter manual test with migration FF off, did not see major issue, LGTM.
Here is the test run I executed:
https://docs.google.com/spreadsheets/d/1qqwd5qiyHN306tdBjd-e5G3t4czB7q1FdWSVrAg-_Y8/edit#gid=1933521616

**Review by michael-s-molina — APPROVED** [review]
LGTM. Thanks @graceguo-supercat for all the hard work!

**Review by jinghua-qa — APPROVED** [review]
LGTM

**Pull request merged**